### PR TITLE
Use caliper

### PR DESCRIPTION
**Pull request by jonesholger**
WIP: this PR adds Caliper profiling for Spot/Hatchet based Analysis
- [x] add profiler_caliper callback using Spot service : skips epoch 0 by default, but is settable via script
- [x] add Tim's synthetic resnet 
- [x] add adiak metadata keys for filtering (i.e compare across compilers, architectures, num nodes,etc)
- [x] add/normalize caliper marking for .cu so Hatchet trees can be compared across cuda and openmp
- [ ] add support or setup env for mpi, OMPT, CUPTI, nvtx, sampler,trace for additional profiling 
- [ ] compare sampler to hpctoolkit
- [x] use Caliper region_profiling for an autotuning OpenMP example : im2col in epoch 0
- [ ] add Hatchet scripts and Spot template for more automated processing; for CI

### CODE REVIEW

**Review by timmoon10 — COMMENTED** [review]
- Revert changes to application scripts.
- Move Caliper marking to base layer class. This would better encapsulate the Caliper-specific code to a smaller region of the code. Something like:
```c++
CALI_CXX_MARK_SCOPE(this->get_type() + " : fp_compute");
```
in [forward prop](https://github.com/LLNL/lbann/blob/develop/src/layers/data_type_layer.cpp#L84) and [back prop](https://github.com/LLNL/lbann/blob/7e0b4affddb5bf2f3c45a0f8a6b2e4ce60eb8960/src/layers/data_type_layer.cpp#L137).

File: include/lbann/utils/summary_impl.hpp [hunk 1]
```diff
@@ -194,6 +195,9 @@ inline void lbann_summary::reduce_2norm(const std::string tag, const El::Abstrac
 
 template <typename TensorDataType>
 inline auto lbann_summary::local_sum(const El::AbstractMatrix<TensorDataType>& mat) const -> BiggerOf<TensorDataType, float> {
+#ifdef LBANN_HAS_CALIPER
+  CALI_CXX_MARK_FUNCTION;
+#endif
```
Comment (benson31) [inline include/lbann/utils/summary_impl.hpp:200]: I would say that we should define this pattern at a higher level:
```
// Maybe base.hpp, maybe elsewhere
#ifdef LBANN_HAS_CALIPER
#define LBANN_CALI_MARK_FUNC CALI_CXX_MARK_FUNCTION
#else
#define LBANN_CALI_MARK_FUNC
#endif
```
Then,
```suggestion
  LBANN_CALI_MARK_FUNC;
```

**Comment by jonesholger** [comment]
Just for convenience spack package.py diff is here :
diff --git a/var/spack/repos/builtin/packages/lbann/package.py b/var/spack/repos/builtin/packages/lbann/package.py
index e252d35253..fe793bbcae 100644
--- a/var/spack/repos/builtin/packages/lbann/package.py
+++ b/var/spack/repos/builtin/packages/lbann/package.py
@@ -72,6 +72,7 @@ class Lbann(CMakePackage, CudaPackage, ROCmPackage):
     variant('pfe', default=True, description='Python Frontend for generating and launching models')
     variant('boost', default=False, description='Enable callbacks that use Boost libraries')
     variant('asan', default=False, description='Build with support for address-sanitizer')
+    variant('caliper',default=False, description='Build with support for Caliper based profiling')

     # LBANN benefits from high performance linkers, but passing these in as command
     # line options forces the linker flags to unnecessarily propagate to all
@@ -237,6 +238,9 @@ class Lbann(CMakePackage, CudaPackage, ROCmPackage):
     generator = 'Ninja'
     depends_on('ninja', type='build')

+    depends_on('caliper@2.6.0',when='@:0.90,0.102: +caliper')
+    depends_on('caliper@2.6.0 +cuda',when='@:0.90,0.102: +caliper +cuda')
+
     @property
     def common_config_args(self):
         spec = self.spec
@@ -306,6 +310,7 @@ def cmake_args(self):
             '-DLBANN_WITH_UNIT_TESTING:BOOL=%s' % (self.run_tests),
             '-DLBANN_WITH_VISION:BOOL=%s' % ('+vision' in spec),
             '-DLBANN_WITH_VTUNE:BOOL=%s' % ('+vtune' in spec),
+            '-DLBANN_WITH_CALIPER:BOOL=%s' % ('+caliper' in spec),
             '-DLBANN_DATATYPE={0}'.format(spec.variants['dtype'].value),
             '-DCEREAL_DIR={0}'.format(spec['cereal'].prefix),
             # protobuf is included by py-protobuf+cpp

**Comment by jonesholger** [comment]
package.py diff is here for convenience

`diff --git a/var/spack/repos/builtin/packages/lbann/package.py b/var/spack/repos/builtin/packages/lbann/package.py
index e252d35253..fe793bbcae 100644
--- a/var/spack/repos/builtin/packages/lbann/package.py
+++ b/var/spack/repos/builtin/packages/lbann/package.py
@@ -72,6 +72,7 @@ class Lbann(CMakePackage, CudaPackage, ROCmPackage):
     variant('pfe', default=True, description='Python Frontend for generating and launching models')
     variant('boost', default=False, description='Enable callbacks that use Boost libraries')
     variant('asan', default=False, description='Build with support for address-sanitizer')
+    variant('caliper',default=False, description='Build with support for Caliper based profiling')

     # LBANN benefits from high performance linkers, but passing these in as command
     # line options forces the linker flags to unnecessarily propagate to all
@@ -237,6 +238,9 @@ class Lbann(CMakePackage, CudaPackage, ROCmPackage):
     generator = 'Ninja'
     depends_on('ninja', type='build')

+    depends_on('caliper@2.6.0',when='@:0.90,0.102: +caliper')
+    depends_on('caliper@2.6.0 +cuda',when='@:0.90,0.102: +caliper +cuda')
+
     @property
     def common_config_args(self):
         spec = self.spec
@@ -306,6 +310,7 @@ def cmake_args(self):
             '-DLBANN_WITH_UNIT_TESTING:BOOL=%s' % (self.run_tests),
             '-DLBANN_WITH_VISION:BOOL=%s' % ('+vision' in spec),
             '-DLBANN_WITH_VTUNE:BOOL=%s' % ('+vtune' in spec),
+            '-DLBANN_WITH_CALIPER:BOOL=%s' % ('+caliper' in spec),
             '-DLBANN_DATATYPE={0}'.format(spec.variants['dtype'].value),
             '-DCEREAL_DIR={0}'.format(spec['cereal'].prefix),
             # protobuf is included by py-protobuf+cpp`

**Comment by jonesholger** [comment]
spack fork https://github.com/jonesholger/spack.git branch lbann_use_caliper has package.py +caliper spec added

File: src/utils/im2col.cpp [hunk 2]
```diff
@@ -406,48 +452,94 @@ void im2col_2d(const TensorDataType *__restrict__ input_buffer,
...
-            const int output_index = output_row + output_col * output_height;
-
-            // Copy input entry to output entry if valid
-            output_buffer[output_index]
-              = input_pos_valid ? input_buffer[input_index] : TensorDataType(0.);
-
-          }
-        }
+  //get original max threads  // we may not want to reset to older max threads since tuned threads impacts program wide
+  unsigned int omp_max_threads=omp_get_max_threads();
```
Comment (benson31) [inline src/utils/im2col.cpp:456]:
```suggestion
```
This variable is unused.

File: CMakeLists.txt [hunk 3]
```diff
@@ -613,6 +637,16 @@ if (LBANN_HAS_VTUNE)
   target_link_libraries(lbann PUBLIC ${VTUNE_STATIC_LIB})
 endif ()
 
+if(LBANN_HAS_CALIPER)
+  target_link_libraries(lbann PUBLIC adiak)
```
Comment (benson31) [inline CMakeLists.txt:641]:
```suggestion
  target_link_libraries(lbann PUBLIC adiak::adiak)
```

File: src/callbacks/imcomm.cpp [hunk 4]
```diff
@@ -32,6 +32,7 @@
 #include "lbann/utils/exception.hpp"
 #include "lbann/utils/summary_impl.hpp"
 #include "lbann/utils/timer.hpp"
+#include "lbann/utils/profiling.hpp"
```
Comment (benson31) [inline src/callbacks/imcomm.cpp:35]:
```suggestion
```

File: src/layers/activations/cereal_registration/softmax.cpp [hunk 5]
```diff
@@ -25,6 +25,7 @@
 ////////////////////////////////////////////////////////////////////////////////
 #include "lbann/utils/serialize.hpp"
 #include <lbann/layers/activations/softmax.hpp>
+#include "lbann/utils/profiling.hpp"
```
Comment (benson31) [inline src/layers/activations/cereal_registration/softmax.cpp:28]:
```suggestion
```

File: src/proto/factories/callback_factory.cpp [hunk 6]
```diff
@@ -183,6 +184,10 @@ void register_default_builders(factory_type& factory)
                            build_print_statistics_callback_from_pbuf);
   factory.register_builder("CallbackProfiler",
                            build_profiler_callback_from_pbuf);
+//#ifdef LBANN_HAS_CALIPER                           
+  factory.register_builder("CallbackProfilerCaliper",
+                           build_profiler_caliper_callback_from_pbuf);
+//#endif
```
Comment (benson31) [inline src/proto/factories/callback_factory.cpp:190]:
```suggestion
  factory.register_builder("CallbackProfilerCaliper",
                           build_profiler_caliper_callback_from_pbuf);
```

File: applications/vision/resnet.py [hunk 7]
```diff
@@ -5,7 +5,6 @@
 import lbann.contrib.args
```
Comment (benson31) [inline applications/vision/resnet.py:5]: Please revert all changes to this file.

**Review by benson31 — COMMENTED** [review]
Some quick comments. Still reading the larger bits of code changes.

File: CMakeLists.txt (hunk 3 above)
Comment (jonesholger) [reply CMakeLists.txt:641]: namespace added in Adiak  version 0.2.2 which isn't in Spack preferred 0.2.1
install(EXPORT adiak-targets NAMESPACE adiak:: DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/adiak)

Yes, I ran into this before way back when for hand compiles. I'll look into making this consistent

Comment (benson31) [reply CMakeLists.txt:641]:
```suggestion
  if (TARGET adiak::adiak)
    target_link_libraries(lbann PUBLIC adiak::adiak)
  else ()
    target_link_libraries(lbann PUBLIC adiak)
  endif ()
```
Would this work?

**Review by benson31 — CHANGES_REQUESTED** [review]
In addition to the specific issues mentioned in the comments, I have a few general questions/comments:

First, marking the functions is terribly invasive in terms of the diff. For the layers in particular, why not simply add scope marking to `data_type_layer::forward_prop()` and the corresponding back prop function? Then only those two functions in a single file need to change... Yes, constructing the scope string incurs some (small, CPU) runtime cost, but then every single layer gets annotated, forward and backward prop, "for free".

Second, if we continue individually annotating each and every layer's `{fp,bp}_compute` functions, what's the motivation for when to use `LBANN_CALIPER_MARK_FUNCTION` vs `LBANN_CALIPER_MARK_SCOPE("layer_name::function_name")`? Is there a reason to use one vs the other, and is there a consistent scheme for when to use which?

I'm still reviewing the callback implementation and the im2col changes, but these are just some high-level observations on the other 130 files.

File: CMakeLists.txt [hunk 8]
```diff
@@ -557,6 +560,22 @@ if (LBANN_WITH_VTUNE)
...
+  if(caliper_FOUND AND adiak_FOUND)
+    set(LBANN_HAS_CALIPER TRUE)
+    message(STATUS "LBANN_WITH_CALIPER")
+    configure_file(
+      "${CMAKE_SOURCE_DIR}/cmake/configure_files/adiak_config.hpp.in"
+      "${CMAKE_BINARY_DIR}/adiak_config.hpp"
+      @ONLY)
+  else ()
+    set(LBANN_HAS_CALIPER FALSE)
+    set(LBANN_WITH_CALIPER OFF)
```
Comment (ndryden) [inline CMakeLists.txt:575]: Is this branch meaningful? `find_package(... REQUIRED)` should kill the build process if these are not found.

File: CMakeLists.txt [hunk 9]
```diff
@@ -711,6 +735,20 @@ if (LBANN_HAS_VTUNE)
...
+  else ()
+    target_link_libraries(lbann PUBLIC adiak::adiak)
+  endif ()
+  target_link_libraries(lbann PUBLIC caliper)
+endif ()
+
+if (LBANN_HAS_NVSHMEM)
+  set_property(TARGET lbann PROPERTY CUDA_SEPARABLE_COMPILATION ON)
+  target_link_libraries(lbann PUBLIC NVSHMEM::NVSHMEM)
+endif ()
```
[the remark this answers is not in the document]
Comment (ndryden) [reply CMakeLists.txt:750]: Can we kill this from this PR?

File: cmake/configure_files/adiak_config.hpp.in [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  constexpr static const char* compiler_flags_debug = "@CMAKE_CXX_FLAGS_DEBUG@";
+  constexpr static const char* cuda_compiler_version =
+    "@CMAKE_CUDA_COMPILER_VERSION@";
+  constexpr static const char* cuda_flags = "@CMAKE_CUDA_FLAGS@";
+  constexpr static const char* cuda_flags_release =
+    "@CMAKE_CUDA_FLAGS_RELEASE@";
+  constexpr static const char* cuda_flags_relwithdebinfo =
+    "@CMAKE_CUDA_FLAGS_RELWITHDEBINFO@";
+  constexpr static const char* cuda_flags_debug = "@CMAKE_CUDA_FLAGS_DEBUG@";
+}; // struct adiak_configuration
```
Comment (ndryden) [inline cmake/configure_files/adiak_config.hpp.in:69]: We should probably add the associated HIP flags, too.

File: src/utils/random.cpp [hunk 11]
```diff
@@ -24,13 +24,16 @@
...
-#include <omp.h>
 #define LBANN_RANDOM_INSTANTIATE
 #include "lbann/io/file_io.hpp"
 #include "lbann/utils/hash.hpp"
+#include "lbann/utils/profiling.hpp"
 #include "lbann/utils/random.hpp"
+
 #include <thread>
 
+#include <omp.h>
```
Comment (ndryden) [inline src/utils/random.cpp:35]: Should this be wrapped in a check for `_OPENMP`?

File: src/utils/profiling.cpp [hunk 12]
```diff
@@ -42,17 +42,166 @@
...
+  // todo get lib : e.g libomp,libiomp5,libgomp etc  : parse adiak::libraries
+  // via tool callback note version map only goes to 5.1; revise as needed
+#if defined(_OPENMP)
+  std::unordered_map<unsigned, std::string> map{{200505, "2.5"},
+                                                {200805, "3.0"},
+                                                {201107, "3.1"},
+                                                {201307, "4.0"},
+                                                {201511, "4.5"},
+                                                {201811, "5.0"},
+                                                {202011, "5.1"}};
```
Comment (ndryden) [inline src/utils/profiling.cpp:187]: 5.2 is out. I think it's `202111`, but I am not certain.

File: cmake/configure_files/adiak_config.hpp.in (hunk 10 above)
Comment (ndryden) [reply cmake/configure_files/adiak_config.hpp.in:69]: (This got added between writing this and submitting the review.)

**Review by ndryden — CHANGES_REQUESTED** [review]
Only some minor cleanups needed, I think.

**Review by ndryden — APPROVED** [review]
Looks good to me.

**Review by tbennun — APPROVED** [review]

**Review by benson31 — APPROVED** [review]
There are some odd annotations (`src/utils/random.cpp`, optimizers, etc), but these should be harmless, so I've left them alone.

**Pull request merged**